File: adapters/src/main/java/com/wang/adapters/utils/ViewGroupWrapUtils.java
package com.wang.adapters.utils;

import android.util.DisplayMetrics;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager.widget.ViewPager;
import androidx.viewpager2.widget.ViewPager2;

public class ViewGroupWrapUtils {

    /**
     * 获得child的宽高
     */
    public static void getChildWidthHeight(ViewGroup vg, @NonNull OnChildWidthHeightResultListener listener) {
        vg.post(() -> {
            if (vg.getChildCount() > 0 && (vg.getChildAt(0).getWidth() > 0 || vg.getChildAt(0).getHeight() > 0)) {
                View childAt = vg.getChildAt(0);
                listener.onResult(childAt.getWidth(), childAt.getHeight());
            } else {
                vg.addOnLayoutChangeListener(new View.OnLayoutChangeListener() {
                    @Override
                    public void onLayoutChange(View v, int left, int top, int right, int bottom, int oldLeft, int oldTop, int oldRight, int oldBottom) {
                        if (vg.getChildCount() > 0) {
                            View childAt = vg.getChildAt(0);
                            if (childAt.getWidth() > 0 || childAt.getHeight() > 0) {
                                vg.removeOnLayoutChangeListener(this);
                                listener.onResult(childAt.getWidth(), childAt.getHeight());
                            }
                        }
                    }
                });
            }
        });
    }

    /**
     * 将vp或rv的高/宽与第一个child高/宽一致（相当于wrap）
     *
     * @param isWidth 修改宽还是高
     */
    public static void wrap(RecyclerView rv, boolean isWidth) {
        rv.post(() -> {
            if (rv.getChildCount() > 0) {
                setMeasureSize(rv, rv.getChildAt(0), isWidth);
            } else {
                rv.addOnChildAttachStateChangeListener(new RecyclerView.OnChildAttachStateChangeListener() {
                    @Override
                    public void onChildViewAttachedToWindow(@NonNull View view) {
                        if (rv.getChildCount() > 0) {
                            rv.removeOnChildAttachStateChangeListener(this);
                            setMeasureSize(rv, rv.getChildAt(0), isWidth);
                        }
                    }

                    @Override
                    public void onChildViewDetachedFromWindow(@NonNull View view) {
                    }
                });
            }
        });
    }

    /**
     * 将vp或rv的高/宽与第一个child高/宽一致（相当于wrap）
     *
     * @param isWidth 修改宽还是高
     */
    public static void wrap(ViewPager vp, boolean isWidth) {
        vp.post(() -> {
            if (vp.getChildCount() > 0) {
                setMeasureSize(vp, vp.getChildAt(0), isWidth);
            } else {
                vp.addOnLayoutChangeListener(new View.OnLayoutChangeListener() {
                    @Override
                    public void onLayoutChange(View v, int left, int top, int right, int bottom, int oldLeft, int oldTop, int oldRight, int oldBottom) {
                        if (vp.getChildCount() > 0) {
                            vp.removeOnLayoutChangeListener(this);
                            setMeasureSize(vp, vp.getChildAt(0), isWidth);
                        }
                    }
                });
            }
        });
    }

    /**
     * 将vp或rv的高/宽与第一个child高/宽一致（相当于wrap）
     *
     * @param isWidth 修改宽还是高
     */
    public static void wrap(ViewPager2 vp, boolean isWidth) {
        RecyclerView rv = (RecyclerView) vp.getChildAt(0);
        rv.post(() -> {
            if (rv.getChildCount() > 0) {
                setMeasureSize(vp, rv.getChildAt(0), isWidth);
            } else {
                rv.addOnChildAttachStateChangeListener(new RecyclerView.OnChildAttachStateChangeListener() {
                    @Override
                    public void onChildViewAttachedToWindow(@NonNull View view) {
                        if (rv.getChildCount() > 0) {
                            rv.removeOnChildAttachStateChangeListener(this);
                            setMeasureSize(vp, rv.getChildAt(0), isWidth);
                        }
                    }

                    @Override
                    public void onChildViewDetachedFromWindow(@NonNull View view) {
                    }
                });
            }
        });
    }

    private static void setMeasureSize(ViewGroup vg, View child, boolean isWidth) {
        child.post(() -> {
            DisplayMetrics dm = vg.getResources().getDisplayMetrics();
            if (isWidth) {
                int screenWidth = dm.widthPixels;
                child.measure(View.MeasureSpec.makeMeasureSpec(screenWidth * 5, View.MeasureSpec.AT_MOST),
                        View.MeasureSpec.makeMeasureSpec(child.getMeasuredHeight(), View.MeasureSpec.EXACTLY));

                vg.getLayoutParams().width = child.getMeasuredWidth() + vg.getPaddingLeft() + vg.getPaddingRight();
            } else {
                int screenHeight = dm.heightPixels;
                child.measure(View.MeasureSpec.makeMeasureSpec(child.getMeasuredWidth(), View.MeasureSpec.EXACTLY),
                        View.MeasureSpec.makeMeasureSpec(screenHeight * 5, View.MeasureSpec.AT_MOST));

                vg.getLayoutParams().height = child.getMeasuredHeight() + vg.getPaddingTop() + vg.getPaddingBottom();
            }
            vg.setLayoutParams(vg.getLayoutParams());
        });
    }

    public interface OnChildWidthHeightResultListener {
        void onResult(int width, int height);
    }
}

File: adapters/src/main/java/com/wang/adapters/utils/ViewGroupWrapUtils.kt
package com.wang.adapters.utils

import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager.widget.ViewPager
import androidx.viewpager2.widget.ViewPager2

object ViewGroupWrapUtils {
    /**
     * 获得child的宽高
     */
    fun getChildWidthHeight(vg: ViewGroup, listener: OnChildWidthHeightResultListener) {
        vg.post {
            if (vg.childCount > 0 && (vg.getChildAt(0).width > 0 || vg.getChildAt(0).height > 0)) {
                val childAt = vg.getChildAt(0)
                listener.onResult(childAt.width, childAt.height)
            } else {
                vg.addOnLayoutChangeListener(object : View.OnLayoutChangeListener {
                    override fun onLayoutChange(v: View, left: Int, top: Int, right: Int, bottom: Int, oldLeft: Int, oldTop: Int, oldRight: Int, oldBottom: Int) {
                        if (vg.childCount > 0) {
                            val childAt = vg.getChildAt(0)
                            if (childAt.width > 0 || childAt.height > 0) {
                                vg.removeOnLayoutChangeListener(this)
                                listener.onResult(childAt.width, childAt.height)
                            }
                        }
                    }
                })
            }
        }
    }

    /**
     * 将vp或rv的高/宽与第一个child高/宽一致（相当于wrap）
     *
     * @param isWidth 修改宽还是高
     */
    fun wrap(rv: RecyclerView, isWidth: Boolean) {
        rv.post {
            if (rv.childCount > 0) {
                setMeasureSize(rv, rv.getChildAt(0), isWidth)
            } else {
                rv.addOnChildAttachStateChangeListener(object : RecyclerView.OnChildAttachStateChangeListener {
                    override fun onChildViewAttachedToWindow(view: View) {
                        if (rv.childCount > 0) {
                            rv.removeOnChildAttachStateChangeListener(this)
                            setMeasureSize(rv, rv.getChildAt(0), isWidth)
                        }
                    }

                    override fun onChildViewDetachedFromWindow(view: View) {}
                })
            }
        }
    }

    /**
     * 将vp或rv的高/宽与第一个child高/宽一致（相当于wrap）
     *
     * @param isWidth 修改宽还是高
     */
    fun wrap(vp: ViewPager, isWidth: Boolean) {
        vp.post {
            if (vp.childCount > 0) {
                setMeasureSize(vp, vp.getChildAt(0), isWidth)
            } else {
                vp.addOnLayoutChangeListener(object : View.OnLayoutChangeListener {
                    override fun onLayoutChange(v: View, left: Int, top: Int, right: Int, bottom: Int, oldLeft: Int, oldTop: Int, oldRight: Int, oldBottom: Int) {
                        if (vp.childCount > 0) {
                            vp.removeOnLayoutChangeListener(this)
                            setMeasureSize(vp, vp.getChildAt(0), isWidth)
                        }
                    }
                })
            }
        }
    }

    /**
     * 将vp或rv的高/宽与第一个child高/宽一致（相当于wrap）
     *
     * @param isWidth 修改宽还是高
     */
    fun wrap(vp: ViewPager2, isWidth: Boolean) {
        val rv = vp.getChildAt(0) as RecyclerView
        rv.post {
            if (rv.childCount > 0) {
                setMeasureSize(vp, rv.getChildAt(0), isWidth)
            } else {
                rv.addOnChildAttachStateChangeListener(object : RecyclerView.OnChildAttachStateChangeListener {
                    override fun onChildViewAttachedToWindow(view: View) {
                        if (rv.childCount > 0) {
                            rv.removeOnChildAttachStateChangeListener(this)
                            setMeasureSize(vp, rv.getChildAt(0), isWidth)
                        }
                    }

                    override fun onChildViewDetachedFromWindow(view: View) {}
                })
            }
        }
    }

    private fun setMeasureSize(vg: ViewGroup, child: View, isWidth: Boolean) {
        child.post {
            val dm = vg.resources.displayMetrics
            if (isWidth) {
                val screenWidth = dm.widthPixels
                child.measure(
                    View.MeasureSpec.makeMeasureSpec(screenWidth * 5, View.MeasureSpec.AT_MOST),
                    View.MeasureSpec.makeMeasureSpec(child.measuredHeight, View.MeasureSpec.EXACTLY)
                )
                vg.layoutParams.width = child.measuredWidth + vg.paddingLeft + vg.paddingRight
            } else {
                val screenHeight = dm.heightPixels
                child.measure(
                    View.MeasureSpec.makeMeasureSpec(child.measuredWidth, View.MeasureSpec.EXACTLY),
                    View.MeasureSpec.makeMeasureSpec(screenHeight * 5, View.MeasureSpec.AT_MOST)
                )
                vg.layoutParams.height = child.measuredHeight + vg.paddingTop + vg.paddingBottom
            }
            vg.layoutParams = vg.layoutParams
        }
    }
}